app/invocation/invocation_action_card.tsx

```diff
@@ -319,122 +319,161 @@ export default class InvocationActionCardComponent extends React.Component<Props
-        timestamp: metadata?.workerStartTimestamp,
+        timestamp: metadata.workerStartTimestamp,
       },
       {
         name: "Downloading inputs",
         color: "#FF6F00",
-        timestamp: metadata?.inputFetchStartTimestamp,
+        timestamp: metadata.inputFetchStartTimestamp,
       },
       {
         name: "Preparing runner",
         color: "#673AB7",
-        timestamp: metadata?.inputFetchCompletedTimestamp,
+        timestamp: metadata.inputFetchCompletedTimestamp,
       },
       {
         name: "Executing",
         color: "#1E88E5",
-        timestamp: metadata?.executionStartTimestamp,
+        timestamp: metadata.executionStartTimestamp,
       },
       {
         name: "Preparing for upload",
         color: "#673AB7",
-        timestamp: metadata?.executionCompletedTimestamp,
+        timestamp: metadata.executionCompletedTimestamp,
       },
       {
         name: "Uploading outputs",
         color: "#FF6F00",
-        timestamp: metadata?.outputUploadStartTimestamp,
+        timestamp: metadata.outputUploadStartTimestamp,
       },
       // End marker -- not actually rendered.
       {
         name: "Upload complete",
         color: "",
-        timestamp: metadata?.outputUploadCompletedTimestamp,
+        timestamp: metadata.outputUploadCompletedTimestamp,
       },
     ];
 
-    const filteredEvents = events.filter((event) => event && event.timestamp);
-    if (filteredEvents.length == 0) {
+    type FilteredTimelineEvent = {
+      name: string;
+      color: string;
+      timestamp: Timestamp;
+    };
+    let filteredEvents: FilteredTimelineEvent[] = events.filter(
+      (event): event is FilteredTimelineEvent => event.timestamp != null
+    );
+
+    if (filteredEvents.length <= 1) {
       return null;
     }
-    const startTimestamp = filteredEvents[0].timestamp;
 
-    const totalDuration = durationSeconds(
-      filteredEvents[0].timestamp!,
-      filteredEvents[filteredEvents.length - 1].timestamp!
-    );
+    filteredEvents = filteredEvents.map((event, i) => {
+      if (i == filteredEvents.length - 1) return event;
+      const next = filteredEvents[i + 1];
+      // Queued event is recorded by the scheduler (not the workers)
+      // which may result in confusing timestamps due to clock drift.
+      // To reduce confusion, apply an adjustment here to enforce
+      // that all timestamps appear in monotonically increasing order
+      // when rendered as a timeline.
+      event.timestamp =
+        timestampToUnixSeconds(event.timestamp) > timestampToUnixSeconds(next.timestamp)
+          ? next.timestamp
+          : event.timestamp;
+
+      return event;
+    });
+
+    const allUnixTimestamp = filteredEvents.map((e) => timestampToUnixSeconds(e.timestamp));
+    const startUnixTimestamp = Math.min(...allUnixTimestamp);
+    const endUnixTimestamp = Math.max(...allUnixTimestamp);
+    const totalDuration = Math.max(0, endUnixTimestamp - startUnixTimestamp);
```

Comment:
I think this could be reverted to the old logic now that the timestamps are guaranteed to be monotonically increasing

Reply:
Will revert to old logic